### PR TITLE
Serialization support for all circuit components

### PR DESCRIPTION
**Pull request by Nashtare**
This PR introduces serialization support for gates, generators, and circuit component (with primary goal here to support `AllRecursiveCircuits` serialization).

The approach builds over #670, reusing the same approach for gate serialization and a similar one for generator serialization. The `DummyProofGenerator` is a specific one which requires different handling, with associated documentation in the serializer macro description.

It doesn't address the suggestion in #585 yet, hence the specific `to_buffer` / `from_buffer` methods on the evm/ side for the internals of `AllRecursiveCircuits` to remove the copies. I assumed it could be handled separately on a different PR.

The PR also includes an additional serialization check on the `square_root` example, which introduces a new generator outside of the plonky2 crate, to highlight flexibility for projects using plonky2 as backend but with a need for additional gates / generators.

@Sladuca given the work on the gate serialization was originally yours, I took the liberty to mention you as co-author of commit 8b6f75eafb9951743415193a6ad951630a961992, let me know if that's ok for you.

### CODE REVIEW

File: plonky2/src/iop/generator.rs [hunk 1]
```diff
@@ -106,6 +110,52 @@ pub trait WitnessGenerator<F: Field>: 'static + Send + Sync + Debug {
...
+impl<F: Field> PartialEq for WitnessGeneratorRef<F> {
+    fn eq(&self, other: &Self) -> bool {
+        let mut buf1 = Vec::new();
+        let mut buf2 = Vec::new();
+        self.0.serialize(&mut buf1).unwrap();
+        other.0.serialize(&mut buf2).unwrap();
+
+        buf1 == buf2
+    }
+}
```
Comment (Nashtare) [inline plonky2/src/iop/generator.rs:149]: Definitely not a fan of this. We could alternatively just check that both "id" match, but that would only enforce that we use the same ordering of generators. Maybe that would be acceptable?

File: plonky2/src/gates/gate.rs [hunk 2]
```diff
@@ -191,9 +198,19 @@ pub trait Gate<F: RichField + Extendable<D>, const D: usize>: 'static + Send + S
     }
 }
 
+pub trait AnyGate<F: RichField + Extendable<D>, const D: usize>: Gate<F, D> {
```
Comment (npwardberkeley) [inline plonky2/src/gates/gate.rs:201]: This is just to allow gate type checking in `get_gate_tag_impl`, right? I think it'd be nice to have a comment explaining that

File: plonky2/src/iop/generator.rs [hunk 3]
```diff
@@ -106,6 +110,52 @@ pub trait WitnessGenerator<F: Field>: 'static + Send + Sync + Debug {
...
     fn run(&self, witness: &PartitionWitness<F>, out_buffer: &mut GeneratedValues<F>) -> bool;
+
+    fn serialize(&self, dst: &mut Vec<u8>) -> IoResult<()>;
+
+    fn deserialize(src: &mut Buffer) -> IoResult<Self>
+    where
+        Self: Sized;
+}
+
+pub trait AnyWitnessGenerator<F: Field>: WitnessGenerator<F> {
```
Comment (npwardberkeley) [inline plonky2/src/iop/generator.rs:121]: Why do we need this trait, instead of having `WitnessGeneratorRef` wrap `Box<dyn WitnessGenerator>`?

Comment (Nashtare) [reply plonky2/src/iop/generator.rs:121]: That should suffice indeed. Initially I was puzzled by how to deal here the best way, given that it had to handle both WitnessGenerator and SimpleGenerator, but then I fell back to the "easy" hack around of generators ids and Default impl to get to the SimpleGenerator trait.

Comment (Nashtare) [reply plonky2/src/iop/generator.rs:121]: You're right this one is a leftover from a previous attempt, I can remove it.

File: plonky2/src/iop/generator.rs (hunk 1 above)
Comment (Nashtare) [reply plonky2/src/iop/generator.rs]: Given that checking equality by "id" was already done for gates, I opted for this as well here.

File: plonky2/src/iop/generator.rs (hunk 3 above)
Comment (Nashtare) [reply plonky2/src/iop/generator.rs]: I applied your comments, in case we opt to keep going for an approach like this one.

**Review by npwardberkeley — APPROVED** [review]

**Pull request merged**